pyartcd/pyartcd/pipelines/build_rhcos.py

```diff
@@ -0,0 +1,197 @@
+                for b in self.request_session.get(
+                    f"{JENKINS_BASE_URL}/job/{job}/api/json?tree=builds[number,description,result,actions[parameters[name,value]]]",
+                ).json()["builds"]
+                if b["result"] is None  # build is still running when it has no status
+            )
+
+        return [b for b in builds if b["parameters"].get("STREAM") == self.version]
+
+    def start_build(self):
+        """Start a new build for the given version"""
+        # determine parameters
+        params = dict(STREAM=self.version, EARLY_ARCH_JOBS="false")
+        if self.new_build:
+            params["FORCE"] = "true"
+
+        # start the build
+        self.request_session.post(
+            f"{JENKINS_BASE_URL}/job/build/buildWithParameters",
+            data=params,
+        )
+
+        # wait for a related build to begin
+        initial_builds = []
+        for _ in range(300):
+            if initial_builds:
+                break
+            time.sleep(1)  # may take a few seconds for the build to start
+            initial_builds  = self.query_existing_builds()
+        else:  # only gets here if the for loop reaches the count
+            raise Exception("Waited too long for build to start")
+
+        return initial_builds
+
+    def build_result(self, job, number):
+        """Query the status of a known build"""
+        return next((  # expecting exactly one result
+            dict(
+                url=self.build_url(job, b["number"]),
+                description=b["description"] or "[no description yet]",
+                result=b["result"],
+            ) for b in self.request_session.get(
+                f"{JENKINS_BASE_URL}/job/{job}/api/json?tree=builds[number,description,result]"
+            ).json()["builds"]
+            if b["number"] == number
+        ), None)
+
+    def wait_for_builds(self):
+        """Wait for all builds for this version to complete, and give status updates on stderr"""
+        builds_seen: Dict[Tuple, str] = {}
+        completed_builds: Dict[Tuple, Dict] = {}
+        for _ in range(1440):  # x 10s = about 4 hours (slower if bad/no response)
+            new_builds_seen: Dict[Tuple, str] = {
+                (b["job"], b["number"]): b["description"] or "[no description yet]"
+                for b in self.query_existing_builds()
+            }
+
+            # check if any previous builds have newly completed
+            for spec in builds_seen.keys():
+                if spec not in new_builds_seen and spec not in completed_builds:
+                    completed_builds[spec] = completed = self.build_result(*spec)
+                    if completed:  # silently ignore if it's somehow not there... should never happen
+                        print(f"{completed['url']} finished with {completed['result']}: '{completed['description']}'", file=sys.stderr)
+
+            # if there are no builds left running, we're done
+            if not new_builds_seen:
```

Comment:
Wouldn't this break even if there are pending builds in builds_seen?

Reply:
yes, that's why `start_build` waits until a build shows up before it considers its job complete (entirely possible that jenkins could be out of executors or dies or whatever and the build is delayed or lost).

Reply:
I missed that new_builds_seen is querying existing builds in each run - so makes sense now